src/device/net.rs

```diff
@@ -77,32 +185,59 @@ impl<H: Hal, T: Transport> VirtIONet<H, T> {
         self.recv_queue.can_pop()
     }
 
-    /// Receive a packet.
-    pub fn recv(&mut self, buf: &mut [u8]) -> Result<usize> {
-        let mut header = MaybeUninit::<Header>::uninit();
-        let header_buf = unsafe { (*header.as_mut_ptr()).as_bytes_mut() };
-        let len = self.recv_queue.add_notify_wait_pop(
-            &[],
-            &mut [header_buf, buf],
-            &mut self.transport,
-        )?;
-        // let header = unsafe { header.assume_init() };
-        Ok(len as usize - size_of::<Header>())
+    /// Receives a [`RxBuffer`] from network. If currently no data, returns an
+    /// error with type [`Error::NotReady`].
+    ///
+    /// It will try to pop a buffer that completed data reception in the
+    /// NIC queue.
+    pub fn receive(&mut self) -> Result<RxBuffer> {
+        if let Some(token) = self.recv_queue.peek_used() {
+            let mut rx_buf = self.rx_buffers[token as usize].take().unwrap();
```

Comment:
Rather than panicking if the buffer is `None`, which could happen if the device implementation is buggy, how about returning `Err(Error::WrongToken)`?

Reply:
It never happens, when we add a buffer the the queue, it's index is equal to the token.

Reply:
I think what you want here and for the assertion below is actually not `token` but `token % QUEUE_SIZE`, otherwise this is going to crash after the first `QUEUE_SIZE` packets.

However, that aside, it's still possible for a buggy device to mark an invalid buffer as the next used one. Suppose that the device adds buffer 0 to the used ring, then immediately adds the same buffer to the used ring again before the driver marks it as available. It shouldn't do so, it is violating the VirtIO specification, but we shouldn't make our driver crash just because a device is buggy. It's better to return an error so we can try again or disable the device or something.

Reply:
`token` is obtained from `VirtQueue::free_head`, which is obtained from `Descriptor::next`, which is never greater then or equal to `QUEUE_SIZE`.

Reply:
Ah you're right, sorry, the `% QUEUE_SIZE` isn't necessary. We probably should check that the device has sent a valid index though, and return an error rather than panicking if it is out of bounds.